src/ProductManifold.jl

```diff
@@ -0,0 +1,416 @@
+`M_1 × M_2 × M_3`.
+"""
+struct ProductManifold{TM<:Tuple} <: Manifold
+    manifolds::TM
+end
+
+ProductManifold(manifolds::Manifold...) = ProductManifold{typeof(manifolds)}(manifolds)
+
+function cross(M1::Manifold, M2::Manifold)
+    return ProductManifold(M1, M2)
+end
+
+function cross(M1::ProductManifold, M2::Manifold)
+    return ProductManifold(M1.manifolds..., M2)
+end
+
+function cross(M1::Manifold, M2::ProductManifold)
+    return ProductManifold(M1, M2.manifolds...)
+end
+
+function cross(M1::ProductManifold, M2::ProductManifold)
+    return ProductManifold(M1.manifolds..., M2.manifolds...)
+end
+
+"""
+    ShapeSpecification(manifolds::Manifold...)
+
+A structure for specifying array size and offset information for linear
+storage of points and tangent vectors on the product manifold of `manifolds`.
+"""
+struct ShapeSpecification{TRanges, TSizes} end
+
+function ShapeSpecification(manifolds::Manifold...)
+    sizes = map(m -> representation_size(m, MPoint), manifolds)
+    lengths = map(prod, sizes)
+    ranges = UnitRange{Int64}[]
+    k = 1
+    for len ∈ lengths
+        push!(ranges, k:(k+len-1))
+        k += len
+    end
+    TRanges = tuple(ranges...)
+    TSizes = Tuple{map(s -> Tuple{s...}, sizes)...}
+    return ShapeSpecification{TRanges, TSizes}()
+end
+
+"""
+    ProductArray(shape::ShapeSpecification, data)
+
+An array-based representation for points and tangent vectors on the
+product manifold. `data` contains underlying representation of points
+arranged according to `TRanges` and `TSizes` from `shape`.
+Internal views for each specific sub-point are created and stored in `parts`.
+"""
+struct ProductArray{TM<:ShapeSpecification,T,N,TData<:AbstractArray{T,N},TV<:Tuple} <: AbstractArray{T,N}
+    data::TData
+    parts::TV
+end
+
+# The two-argument version of this constructor is substantially faster than
+# the generic one.
+function ProductArray(M::Type{ShapeSpecification{TRanges, Tuple{Size1, Size2}}}, data::TData) where {TRanges, Size1, Size2, T, N, TData<:AbstractArray{T,N}}
+    views = (SizedAbstractArray{Size1}(view(data, TRanges[1])),
+             SizedAbstractArray{Size2}(view(data, TRanges[2])))
+    return ProductArray{M, T, N, TData, typeof(views)}(data, views)
+end
+
+function ProductArray(M::Type{ShapeSpecification{TRanges, Tuple{Size1, Size2, Size3}}}, data::TData) where {TRanges, Size1, Size2, Size3, T, N, TData<:AbstractArray{T,N}}
+    views = (SizedAbstractArray{Size1}(view(data, TRanges[1])),
+             SizedAbstractArray{Size2}(view(data, TRanges[2])),
+             SizedAbstractArray{Size3}(view(data, TRanges[3])))
+    return ProductArray{M, T, N, TData, typeof(views)}(data, views)
+end
+
+function ProductArray(M::ShapeSpecification{TRanges, TSizes}, data::TData) where {TM, TRanges, TSizes, T, N, TData<:AbstractArray{T,N}}
+    return ProductArray(typeof(M), data)
+end
+
+"""
+    prod_point(M::ProductManifold, pts...)
```

Comment:
This docstring doesn't agree with the signature. It's not clear how to use this function.

Reply:
Right, I've improved the documentation.